Add JIT graph fuser for oneDNN Graph API (Preview4)

chunyuan-w opened the pull request:
## Description
Preview4 PR of this [RFC](https://github.com/pytorch/pytorch/issues/49444).

On the basis of https://github.com/pytorch/pytorch/pull/50256, the below improvements are included:

- The [preview4 release branch](https://github.com/oneapi-src/oneDNN/releases/tag/graph-v0.4.1) of the oneDNN Graph API is used
- The fuser now works with the profiling graph executor. We have inserted type check nodes to guard the profiled tensor properties.

### User API:
The optimization pass is disabled by default. Users could enable it by:
```
torch.jit.enable_onednn_fusion(True)
```

### Performance:
[pytorch/benchmark](https://github.com/pytorch/benchmark) tool is used to compare the performance:
- SkyLake 8180 (1 socket of 28 cores):

  ![image](https://user-images.githubusercontent.com/65992142/151162305-05e44425-a24e-4d5e-94e1-743b40b87a8c.png)

- SkyLake 8180 (single thread):

  ![image](https://user-images.githubusercontent.com/65992142/151162528-69f90b79-d08d-46b8-8775-d80a6ccbce8a.png)
 \* By mapping hardswish to oneDNN Graph, it’s 8% faster than PyTorch JIT (NNC + OFI)
  \** We expect performance gain after mapping transpose, contiguous & view to oneDNN graph ops


### Directory structure of the integration code
Fuser-related code are placed under:
```
torch/csrc/jit/codegen/onednn/
```

Optimization pass registration is done in:
```
torch/csrc/jit/passes/onednn_graph_fuser.h
```

CMake for the integration code is:
```
caffe2/CMakeLists.txt
```

## Limitations

- In this PR, we have only supported the optimization on Linux platform. The support on Windows and MacOS will be enabled as the next step.
- We have only optimized the inference use case.

chunyuan-w committed:
add jit symbols

chunyuan-w committed:
add cmake

chunyuan-w committed:
add tests

chunyuan-w committed:
add README

chunyuan-w committed:
add src

chunyuan-w committed:
switch mkl-dnn to dev-graph-preview2

update backend to use oneDNN v2.3.3

sanchitintel committed:
Fix non-MKLDNN builds & lint

Add mkldnn fix

Fix typo

Fix typo again :(

Fix Windows builds

Try to fix MacOS build

Windows build fixed

sanchitintel committed:
Refactor unit-tests & fix flake8 and lint

Disable LLGA build for MacOS & Windows

Configure Bazel build for MKLDNN

Fix clang-format

Fix formatting

Skip lint for const_cast

sanchitintel committed:
Don't compile LlgaTensorImpl.cpp 

A separate Bazel build file needs to be created for LLGA

chunyuan-w committed:
use decorator for oneDNN fusion enabling

chunyuan-w committed:
use C10_EXPORT

Jianhui-Li commented:
@vitaly-fedyunin @gottbrath  FYI that this is the oneDNN Graph API integration. It depends on the https://github.com/pytorch/pytorch/pull/63748.

chunyuan-w committed:
rename files

chunyuan-w committed:
update README

chunyuan-w committed:
fix IR name exposed to user

chunyuan-w committed:
fix constant inputs

chunyuan-w committed:
keep infer_shape for preview2

chunyuan-w committed:
fix clang-tidy

sanchitintel committed:
Move LlgaTensorImpl.h & LlgaTensorImpl.cpp

sanchitintel committed:
Use C++14 primitives for locking

sanchitintel committed:
Remove redundant code

sanchitintel committed:
Fix lint

sanchitintel commented:
CI failures are currently being caused by some issues in the CI infra, and are also occurring with other PRs.

pinzhenx commented on torch/csrc/jit/codegen/onednn/kernel.cpp:
```diff
@@ -0,0 +1,258 @@
+  });
+
+  lock_read();
+  if (is_initialized_) {
```
one minor: just wondering if it's possible to use a standard mutex here, which is more cross-platform.

```
  if (!is_initialized_) {
    std::lock_guard<std::mutex> lock(rw_mutex_);
    if (!is_initialized_) {
      ...
      is_initialized_ = true;
    }
  }

  private:
    std::mutex rw_mutex_;
 ```

sanchitintel replied:
Thanks for your inputs, @pinzhenx!
@chunyuan-w, I took another look & noticed that the custom `lock_read` & `lock_write` primitives are only being used here, so I'm confused why we need to use a read-write lock here?

The Reader using the read-lock doesn't seem to be reading anything -
```
lock_read();
  if (is_initialized_) {
    unlock_read();
  }
```

sanchitintel replied:
Thanks! I noticed that the custom RW lock is only being used here, but it's not actually needed here,
so I'll get rid of it.

pinzhenx replied:
Uh, example above is just describing an idea of portable code. Double checked lock like this isn’t robust actually. It should be further combined with std::atomic (even specify memory order on non-x86). Other than this hand written way, a modern while portable approach could be achieved by std::call_once here.

sanchitintel replied:
Thanks, @pinzhenx. I'll take a look at what exactly the lock is being used for, and will make changes accordingly.

Also, I think atomics aren't required here for correctness because even if a reader were to read a stale value, it'd definitely read the correct one after acquiring the lock because there can be only one writer, which only writes after acquiring the lock. Atomics slow things down, so I'd rather not use it.

However, the current snippet has two differences from the previous version, due to which I was considering revising it anyway:
1. In the previous version, readers would've waited for the writer to finish writing. (I'll see what exactly the lock is being used for, and revisit this decision).
2. In this version, each reader would've to wait on `std::mutex`, which is a performance-bottleneck, so I'll use a RW mutex here.

chunyuan-w replied:
cc @jgong5

pinzhenx commented on BUILD.bazel:
```diff
@@ -318,7 +318,8 @@ filegroup(
     name = "aten_native_mkldnn_cpp",
-    srcs = glob(["aten/src/ATen/native/mkldnn/*.cpp"]),
+    srcs = glob(["aten/src/ATen/native/mkldnn/*.cpp"],
+                exclude = ["aten/src/ATen/native/mkldnn/LlgaTensorImpl.cpp"]),
```
no need for this `exclude`

sanchitintel replied:
Thanks, will remove it, now that the file's been moved.

sanchitintel commented:
CI failures are unrelated.

sanchitintel committed:
Use std::call_once instead of mutex & change ideep commit to latest

sanchitintel committed:
Fix lint

sanchitintel committed:
Merge branch 'master' into chunyuan/llga_preview2

sanchitintel committed:
Delete unused libdnnl_graph.so generated by oneDNN Graph Preview2

sanchitintel committed:
Fix FindMKLDNN.cmake

sanchitintel committed:
[Style] GitHub website is unable to recognize a CMake multi-line comment-block

sanchitintel committed:
Fix infinite recursion bug caught by clang

sanchitintel commented:
The CI failure is unrelated.

sanchitintel committed:
Update README

sanchitintel committed:
Merge branch 'master' into chunyuan/llga_preview2

sanchitintel committed:
Fix lint

eellison commented:
Hi, thank you for the PR!

Do you mind running a larger amount of torchbench and reporting numbers ? You can look at [Jason's post](https://dev-discuss.pytorch.org/t/torchdynamo-update-3-gpu-inference-edition/460) here for what models are supported in script. Initially just the vision models would be useful. @Krovatkin also did some benchmarking of a traced Bert model and found on average a ~16% speedup with this PR.

sanchitintel commented:
Thanks a lot for reviewing, @eellison & @Krovatkin!
We just wanted to let you know that we're working on the benchmarking & will get back to you in a day, or two.

BTW, BERT's (we tested HuggingFace BERT with PyTorch) inference speedup is even higher with oneDNN Graph Preview4 (this PR currently has Preview2, but Preview4 would be available in a few days), as it has MHA fusion support! :)

sanchitintel committed:
Add Preview4 support

sanchitintel committed:
Merge latest commits from the PyTorch master branch

sanchitintel committed:
Make everything compatible

sanchitintel committed:
Add an API to enable/disable oneDNN Graph weight-caching

sanchitintel committed:
Add necessary changes

sanchitintel committed:
Fix CI errors

sanchitintel committed:
Fix clang-tidy

sanchitintel committed:
Remove extra whitespace :(

sanchitintel committed:
Fix performance-unnecessary-copy-initialization warning

sanchitintel committed:
Try to fix build & import errors on CI

sanchitintel committed:
Add the weight-cache API later

sanchitintel committed:
Remove whitespace yet again

sanchitintel committed:
Build static libs on PyTorch CI until conda issue is fixed

On conda, users experience an issue `ImportError: libdnnl_graph.so.0: cannot open shared object file: No such file or directory`.
Will have to see how PyTorch CMake handles other shared libraries.

sanchitintel commented:
Hello @eellison,

We used this [TorchBench branch](https://github.com/sanchitintel/benchmark/tree/onednn-graph-preview2) for comparison. `compare_llga.sh` can be run for comparison.

For benchmarking `mobilenet_v3_large` with `hardswish` support in oneDNN Graph, [this oneDNN Graph branch](https://github.com/sanchitintel/oneDNN/tree/hardswish) can be used in `third_party/ideep/mkl-dnn`. It delivers a speedup over PyTorch JIT (NNC + OFI) because 21 additional reorders are prevented (the major factor here), and fusion with conv also helped further.
The next release of oneDNN Graph would have `hardswish` support.

We're also exploring adding a `hardsigmoid` op in oneDNN Graph.

Thank you!

sanchitintel commented:
Please note that this PR should be merged after #71546, as #71546 changes the  `third_party/ideep` commit (this PR also uses that ideep commit, but it'd probably be better to merge #71546 first, so that oneDNN v2.5.2 upgrade would be in a separate PR). Thank you!

sanchitintel committed:
Merge branch 'master' into chunyuan/llga_preview2

eellison commented on torch/csrc/jit/codegen/onednn/graph_helper.cpp:
```diff
@@ -0,0 +1,521 @@
+  if (!shouldConsiderForMerge(toMerge)) {
+    return false;
+  }
+  return opToOwningPartition_.get(toMerge) ==
```
`opToOwningPartition_.get(toMerge) == opToOwningPartition_.get(subgraph);` what is the purpose here ? don't we want to add new nodes to the subgraph that aren't already in the subgraph?

sanchitintel replied:
`opToOwningPartition_` maps nodes^ to their respective oneDNN Graph partition ID.
We'd merge nodes if oneDNN Graph had determined that they would be in the same oneDNN Graph partition.

^ - Since PyTorch is unsupported on 32 bit platforms, it's safe to use pointers of nodes for the purpose.

eellison commented on torch/csrc/jit/codegen/onednn/graph_helper.cpp:
```diff
@@ -0,0 +1,521 @@
+  }
+
+  GRAPH_DEBUG("Get Partitions");
+  std::vector<dnnl::graph::partition> partitions = g.get_partitions(policy);
```
Do we need to create the partially initially before fusing ? Why not fuse and then create partitions at the end ? would this get rid of  needing to maintain the partition map ?

sanchitintel replied:
Sorry, I should've added this in the comments - The oneDNN Graph library determines which ops would be in the same partition, so we need to create partitions to determine what ops ended up in the same partition. Currently, we've a 2-stage partition-building approach. 

When we first create the oneDNN Graph partitions during `CreateLlgaSubgraphs`, `GraphRewriter`'s `LlgaGraphHelper` object creates a partition map that is not maintained after the `CreateLlgaSubgraphs` pass. 
To maintain this map (and the partitions) for later use, we could probably make it a member of the graph instead, so that they could be freed after the graph would be destructed, but that'd complicate things. We welcome your suggestions  to improve this design.

Now, since we don't maintain the partition-map (or the partitions) created in the first step, we re-create them while creating fusion groups (`prim::oneDNNFusionGroup` are created with `createLlgaKernel`).

eellison commented on torch/csrc/jit/codegen/onednn/guard_shape.cpp:
```diff
@@ -0,0 +1,147 @@
+using tensor_type_converter_t =
+    c10::function_ref<TensorTypePtr(const TensorTypePtr& t)>;
+
+void insertTypeGuardForFusionGroup(
```
Could we just use https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/tensorexpr_fuser.cpp directly ? what is preventing that

sanchitintel replied:
This code is identical to `insertTypeGuard`, except that it also skips guarding inputs that are outputs of `prim::oneDNNFusionGroup`. 

So, can we add an `if` condition to `insertTypeGuard` pertaining to `prim::oneDNNFusionGroup` [here](https://github.com/pytorch/pytorch/blob/72cedba6553d6fdacc55ccac0b1b70446c03ef82/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L284-L287)?

sanchitintel replied:
Oh, thanks! I see your point. The input value's node would be `prim::If` and not `prim::FusionGroup` (or in this case `prim::oneDNNFusionGroup`), so no fusion group would directly output a value into another!

I'll reuse existing code, as you advised.

eellison commented on torch/csrc/jit/codegen/onednn/interface.cpp:
```diff
@@ -0,0 +1,145 @@
+        "After RemoveProfileNodesAndSpecializeTypes. Before mutation removal",
+        g);
+
+    RemoveTensorMutation(g);
```
We should pass in a lambda here so that we don't unmutate nodes which are unsupported in OneDNN (and also cpu for when this is enabled more globally)

sanchitintel replied:
Thank you! I'll add something like this -


```
-    RemoveTensorMutation(g);
+    // Don't unmutate nodes which are unsupported in OneDNN Graph
+    RemoveTensorMutation(g,
+        [&](Node* node) {
+          return createOperator(node).kind() != dnnl::graph::op::kind::Wildcard;
+        });
```

sanchitintel replied:
I tried this, which resulted in perf boost for mobilenet_v3_large, mobilenet_v2, resnet50 & timm_efficientnet, but regressions for other Vision models. I'll look into the reason.

<details>

```
    // Don't remove tensor-mutation for nodes unsupported by oneDNN Graph.
    // However, since currently, we've disabled single-op partitions for
    // most ops, they'd still end up being replaced by their functional
    // variants, because at this point, we've no way to tell if they'd
    // end up in single-op partitions. Maybe we can add another pass to
    // undo unnecessary inplace -> functional equivalent changes after
    // we'd know which ops, although supported by oneDNN Graph,
    // wouldn't be handled by it.
    RemoveTensorMutation(g, [](Node* nodeToFunctionalize) {
        static std::unordered_set<Symbol> supportedOps = {
            aten::add_,
            aten::mul_,
            aten::tanh_,
            aten::elu_,
            aten::relu_,
            aten::relu6_,
            aten::gelu_,
            aten::sqrt_,
            aten::sigmoid_,
            aten::hardtanh_,
            aten::abs_,
            aten::square_,
        };
        return supportedOps.count(nodeToFunctionalize->kind()) != 0;
      });

```

![image](https://user-images.githubusercontent.com/90872293/153663041-6e9aa527-b082-4171-8617-d3263a7ac605.png)

</details>

sanchitintel replied:
> Were there any fusions that occurred in OneDNN that were happening before and now are not?

Since those ops are not mapped to oneDNN Graph, by not unmutating them, no fusion opportunities are lost from the oneDNN Graph perspective.

> It may be that globally unmutating - which NNC and NVFuser don't do yet - may be the best strategy for perf, however it'd be nice to keep changes local to this PR and make that decision a separate if we decide to go with it.

Thanks! Now we're only unmutating nodes that can be mapped to oneDNN Graph.

eellison commented on torch/csrc/jit/codegen/onednn/interface.cpp:
```diff
@@ -0,0 +1,145 @@
+    RemoveTensorMutation(g);
+    RemoveListMutation(g);
+    GRAPH_DUMP("After mutation removal. Before DecomposeOps", g);
+    DecomposeOps(g);
```
We probably don't want to do this here because it's going to remove type annotations from nodes. What decompositions do you need here?

sanchitintel replied:
Thanks, I'll get back to you soon on this. I think we're calling it just because `torch::jit::runNondiffOptimization()` & `torch::jit::ProfilingGraphExecutorImpl::runProfilingInsensitiveOptimizations` were used as reference. 😄

eellison commented on torch/csrc/jit/codegen/onednn/interface.cpp:
```diff
@@ -0,0 +1,145 @@
+    std::vector<TypePtr> types = node->tys(attr::types);
+    const auto num_inputs = types.size();
+
+    GRAPH_DEBUG("num_inputs to guard: ", num_inputs);
```
You probably don't want the DEBUGs running on each invocation, because we will have to do an environment read to check if it's enabled, which will be slow. I would at least guard this with 
```
#if DEBUG
GRAPH_DEBUG
#endif
```
something like that, or check for the environment outside of the op invocation and pass it in so we don't read environment on every run

sanchitintel replied:
Thanks, this is a very good idea! I'll add this change later today.

eellison commented on torch/csrc/jit/codegen/onednn/interface.cpp:
```diff
@@ -0,0 +1,145 @@
+      const c10::TensorTypePtr& guard_tensor_type =
+          types[i]->cast<TensorType>();
+
+      if (!input.isTensor()) {
```
Are never going to have scalar inputs ?

sanchitintel replied:
Most scalar inputs would be `prim::Constant` and would thus be unguarded.
Also, some scalar inputs generated at runtime would be converted to tensors of size 1.
We might need to revisit it in the future, though, so I'll ensure that I keep track of it.

eellison commented on torch/csrc/jit/codegen/onednn/interface.cpp:
```diff
@@ -0,0 +1,145 @@
+      // LLGA partition that has passed the check on input shapes.
+      // It is valid to continue here as long as the output shapes from
+      // oneDNN graph partitions are determined by the input shapes.
+      if (tensor.is_mkldnn()) {
```
It seems a little suspect to not check sizes here but i don't know the logic well enough.

sanchitintel replied:
Tensors between two oneDNN Graph partitions can have different (but optimal) layouts (eg. blocked or channels-last), so they're opaque to PyTorch. So, we need not check their shape.

sanchitintel replied:
We wrap an output tensor of a oneDNN graph partition as MKLDNN tensors only if it's an input to another oneDNN partition, but not a PyTorch op that isn't present in any oneDNN graph partition. Their size & dtype would be same as that expected, but their strides may be different (depending on the layout). Currently, the device can only be CPU. We're overriding guard checks for such tensors based on the assumption that size/dtype/device would be correct, as oneDNN Graph would've generated these tensors for use by a oneDNN Graph partition, as oneDNN Graph is already cognizant of the characteristics of the input tensors of the next partition.

eellison commented on torch/csrc/jit/codegen/onednn/interface.cpp:
```diff
@@ -0,0 +1,145 @@
+      }
+    }
+
+    // TODO: check type and return the right flag
```
`TODO: check type and return the right flag` what do you mean by this ?

sanchitintel replied:
Thanks for catching this! This comment is a remnant of a time long past 😄 
I'll remove it.

eellison reviewed (commented):
Looks great !! Left a bunch of comments. I have not got all the way through.  

A few higher level comments:

- We should be using schema instead of node symbol. This is more robust to new schemas being added that we don't support or wrong input types. 

- It'd be nice to defer the partition mapping until after subgraphs are created. It's not clear to me why that's needed during fusion and it would standardize the fusion process with other fusers. 

- Maybe not as part of this PR, but we should add tests with OpInfos like [here](https://github.com/pytorch/pytorch/pull/70465) and test that we don't give incorrect results. It's one thing to not support certain semantics of an operator but we don't want to have hard errors or incorrect semantics

sanchitintel commented on torch/csrc/jit/codegen/onednn/graph_helper.cpp:
```diff
@@ -0,0 +1,521 @@
+    }
+
+    default:
+      return makeWildcardOp(node);
```
That's right!

The motivation behind `wildcard` is to:

 1) make the integration code with PyTorch simpler by just passing every op to the oneDNN Graph library in the `add_op` call - no need to check beforehand whether the op is supported by oneDNN Graph or not. 
 2) The oneDNN Graph library side guarantees to return cycle-free partitions
 
 I'll add [this link ](https://spec.oneapi.io/onednn-graph/latest/programming_model.html)to the README, or better still, add some of this content to it.

sanchitintel commented on torch/csrc/jit/codegen/onednn/graph_helper.cpp:
```diff
@@ -0,0 +1,521 @@
+  if (isLlgaSubgraph(node)) {
+    return true;
+  }
+  if (isViewOp(node)) {
```
Thank you! You're absolutely right that it's dead-code because the `opToOwningPartition_` map wouldn't have this node.
I'll remove it.

sanchitintel commented on torch/csrc/jit/codegen/onednn/graph_helper.cpp:
```diff
@@ -0,0 +1,521 @@
+    try {
+      g.add_op(op);
+    } catch (std::exception& e) {
+      GRAPH_DEBUG(
```
As an example, the oneDNN Graph library has added more strict `dtype` checks during `dnnl::graph::graph::add_op()`. 
Currently, with PyTorch, we only have FP32 support.
The check conditions could be different for different OPs. If the oneDNN Graph backend failed to add an operator, we will add it as a Wildcard instead.

sanchitintel commented on torch/csrc/jit/codegen/onednn/graph_helper.cpp:
```diff
@@ -0,0 +1,521 @@
+  if (!tt)
+    return at::kCPU;
+  auto device = tt->device();
+  if (!device)
```
Currently, oneDNN Graph can only be used with CPU, and users would've to explicitly enable it, so can we get back to it in the future?

sanchitintel commented on torch/csrc/jit/codegen/onednn/graph_helper.cpp:
```diff
@@ -0,0 +1,521 @@
+      auto listConstruct = node->input(0)->node();
+      for (auto input : listConstruct->inputs())
+        o.setInputValue(input);
+      return o.setOutput(0).setAttr("axis", Operator::Int, 1);
```
The input with index 1, as in the second input of node.
On a sidenote, the schema of oneDNN Graph ops has been defined in `third_party/ideep/mkl-dnn/src/interface/op_def.hpp`. I'll mention that in the README file as well, as well as in this functions's (`createOperator`) comments.

sanchitintel commented on torch/csrc/jit/codegen/onednn/graph_helper.cpp:
```diff
@@ -0,0 +1,521 @@
+    }
+
+    case aten::layer_norm: {
+      auto normalized_shape = Operator::Ints(node, 1);
```
Thanks for catching this! Currently, oneDNN (and thus oneDNN Graph) only supports layer norm for the last dim, so the logic here should've been to fall through to the default case if layer norm is to be done on more than 1 dims. I'll fix it.

sanchitintel commented on torch/csrc/jit/codegen/onednn/graph_helper.cpp:
```diff
@@ -0,0 +1,521 @@
+          .setAttr("pads_end", Operator::Ints, 4)
+          .setAttr("dilations", Operator::Ints, 5)
+          .setAttr("groups", Operator::Int, 6)
+          .setAttr("filter_format", std::string("OIX"));
```
Just like `NCHW` is sometimes denoted by `NCX`, OIX means that the shape of the filter tensor is (out_channels, in_channels // groups, spatial_shape). I'll add a comment there. Thanks for pointing this out!

sanchitintel commented on torch/csrc/jit/codegen/onednn/defer_size_check.cpp:
```diff
@@ -0,0 +1,81 @@
+    if (!onlyUsedByShapePreserveOp)
+      return false;
+
+    for (const auto& use : uses) {
```
In some cases, moving `aten::size` can allow more oneDNN Graph fusion opportunities:

While `oneDNN Graph` doesn't actually handle `addmm`, the example in the comments is for illustration -

  
```
    %b = addmm(%a)
    %sz = aten::size(%b)
    %c = relu(%b)
      =>
    %b = addmm(%a)
    %c = relu(%b)
    %sz = aten::size(%c)
```

In the first case, `addmm` & `relu` wouldn't end up in the same oneDNN graph partition, although they can be fused together. They'd be separated by a `wildcard` oneDNN Graph op corresponding to `aten::size`.
In the second case (transformation), `addmm` & `relu` would end up in the same oneDNN Graph partition.

sanchitintel commented on torch/csrc/jit/codegen/onednn/graph_fuser.cpp:
```diff
@@ -0,0 +1,212 @@
+  }
+
+ private:
+  std::vector<WorkBlock> buildWorkBlocks() {
```
We can definitely do so!
 However, please note that currently, `buildWorkBlocks`, `scanNode`, `tryMerge` have the same code-structure as their counterparts in `torch/csrc/jit/passes/create_autodiff_subgraphs.cpp`. Please confirm if you still recommend factoring them out into a common helper file. Thanks!

sanchitintel committed:
Make minor changes before making important changes

sanchitintel commented on torch/csrc/jit/codegen/onednn/defer_size_check.cpp:
```diff
@@ -0,0 +1,81 @@
+        uses.size() > 1 &&
+        std::all_of(uses.begin(), uses.end(), [node](auto& u) {
+          return u.user == node ||
+              // TODO: register more shape preserved op
```
Thanks! In the absence of such an API, would it be okay if I simply call `get_tensorexpr_elementwise_set().getAllKeysAndValues()`? -

```

-        std::all_of(uses.begin(), uses.end(), [node](auto& u) {
-          return u.user == node ||
-              // TODO: register more shape preserved op
-              u.user->matches("aten::relu(Tensor self) -> Tensor") ||
-              u.user->matches("aten::sigmoid(Tensor self) -> Tensor");
-        });
+        std::all_of(uses.begin(), uses.end(),
+          [&](auto& u) {
+            if (u.user == node) {
+              return true;
+            }
+            // match with shape-preserving unary ops in
+            // tensorexpr_elementwise_set that's defined in
+            // torch/csrc/jit/runtime/symbolic_shape_registry_util.cpp
+            // TODO: this matching can perhaps be done more efficiently.
+            auto allSchemas =
+                get_tensorexpr_elementwise_set().getAllKeysAndValues();
+            for (const auto& pair : allSchemas) {
+              const FunctionSchema schemaString = pair.first->schema();
+              const std::string& functionType = pair.second;
+              if ((functionType == "unary") &&
+                  (u.user->matches(schemaString))) {
+                return true;
+              }
+            }
+            return false;
+          });


```

eellison reviewed (commented):
Some more comments..

eellison commented on torch/csrc/jit/codegen/onednn/prepare_binary.cpp:
```diff
@@ -0,0 +1,105 @@
+      EliminateIdentityMulAdd(sub);
+    }
+
+    if ((node->kind() == aten::add && compareConstValue(node->input(1), 0.0)) ||
```
This already exists here but with more rigorous checking of correctness conditions. https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/peephole_alias_sensitive.cpp#L80 Something like this probably makes more sense to do after fusion where we've already guarded that the fusion group is floating point and everything in it is functional

sanchitintel commented on torch/csrc/jit/codegen/onednn/kernel.cpp:
```diff
@@ -0,0 +1,243 @@
+  });
+
+  // Even in case of concurrent threads, the kernel would be initialized once.
+  std::call_once(initialized_flag,
```
If input tensor characteristics would somehow change, then they'd deviate from their corresponding oneDNN Graph's logical tensor, and the partition compilation would fail, resulting in an exception being thrown by oneDNN Graph.

Can you please provide an example of when input tensors might change? I'm guessing you might've encountered such an example. I'll try searching the commit history in the master branch to find such an example. Thanks!

sanchitintel commented on torch/csrc/jit/codegen/onednn/prepare_binary.cpp:
```diff
@@ -0,0 +1,105 @@
+    t->setType(target_type);
+
+    // tensor(42.0) : Float([])  -->  tensor([42.0]) : Float([1])
+    auto unsqueezed = g->insert(aten::unsqueeze, {t, 0});
```
Thanks! I'll check if we'd encounter such a case because currently we only support FP32.

eellison replied:
Yes, the initial integration is for fp32, but eventually, we want this code to always be running, and that might be on models with mixed precision etc. It's one thing to not support things within onednn but we don't want to effect correctness of other code..

Speaking of which... what is the viability of bfloat16 / amp within onednn ? we're making progress on supporting autocasting within JIT so I'm curious there

sanchitintel replied:
> what is the viability of bfloat16 / amp within onednn ? we're making progress on supporting autocasting within JIT so I'm curious there

Sorry, just to clarify, I meant that for our current integration code (bridge between PyTorch & oneDNN Graph), we only support FP32. oneDNN Graph does [support BF16](https://spec.oneapi.io/onednn-graph/latest/ops/lower_precision/TypeCast_1.html) & int8. 

Can I mark it as a TODO for now, and revisit it when we'd add BF16 support in the PyTorch-oneDNN-Graph bridge-code in the near future? Thanks!

sanchitintel reviewed (commented):
Thanks again, @eelison!

I'll make revisions corresponding to your recommendations, and also complete these other tasks -
 - [x] Only unmutating nodes of ops supported by oneDNN Graph.
- [x] Move `aten::size` for more unary ops using `find`.
- [x]  Using `GRAPH_DEBUG` only if a debug macro has been defined, in order to reduce branching & wasteful computation.

sanchitintel commented on torch/csrc/jit/codegen/onednn/graph_helper.cpp:
```diff
@@ -0,0 +1,519 @@
+      return isBetterSuitedForLLGA(kindOfNode);
+    } else {
+      // multi-op partition
+      return true;
```
If a partition has multiple ops supported by oneDNN Graph, then it'd be more efficient for it to be handled by oneDNN Graph. Single-op oneDNN Graph partitions are only used for a few compute-intensive ops, because otherwise, there might be additional reorders, whose overhead might exceed the benefit of letting oneDNN Graph handle them.

sanchitintel committed:
More changes

sanchitintel committed:
Remove try-catch block from LlgaHelper

sanchitintel committed:
GRAPH_DEBUG invocations in the hot-path should be compiled only if the macro GRAPH_DEBUG_ENABLED is true

sanchitintel committed:
More changes

sanchitintel committed:
Fix typo

sanchitintel committed:
Merge branch 'master' into chunyuan/llga_preview2

sanchitintel committed:
Fix clang-format

sanchitintel committed:
Fix format

sanchitintel committed:
Fix formatting

sanchitintel commented on torch/csrc/jit/codegen/onednn/kernel.cpp:
```diff
@@ -0,0 +1,243 @@
+    auto opt = c10::TensorOptions(spec.aten_scalar_type()).device(device_);
+
+    auto outputId = spec.tid();
+    auto iter = inplacePairs_.find(outputId);
```
This is a good point! To avoid these hash-map lookups at runtime, we can add a new member to `LlgaTensorDesc`.

sanchitintel commented on torch/csrc/jit/codegen/onednn/kernel.cpp:
```diff
@@ -0,0 +1,243 @@
+        {spec.logical_tensor(), Engine::getEngine(), input.data_ptr()});
+  }
+  auto numConstantInputs = constantInputs_.size();
+  for (size_t i = 0; i < numConstantInputs; i++) {
```
The storage of constant tensors is being reused in each run. 

As for logical tensors, it's convenient to create logical tensors for each run, because otherwise, memory-management would become complex.

sanchitintel committed:
Precompute some stuff for inference runs

sanchitintel committed:
Fix style

eellison commented on torch/csrc/jit/codegen/onednn/graph_helper.cpp:
```diff
@@ -0,0 +1,561 @@
+// Verify that input tensors are compatible with oneDNN Graph.
+// Scalars would be converted to 1-D tensors later anyway,
+// but they shouldn't be complex-double
+// If this check fails, convert op to wildcard
```
nit: still don't like the fact that we know that this will fail and then still try to send it to OneDNN, it makes debugging why something wasn't failed harder to follow

eellison commented on torch/csrc/jit/codegen/onednn/graph_helper.cpp:
```diff
@@ -0,0 +1,561 @@
+bool checkInputCompatibility(Node* node) {
+  auto allInputs = node->inputs();
+  for (auto input : allInputs) {
+    c10::IValue inputIValue = toIValue(input);
```
`toIValue(input)` only returns anIValue on a constant and (I thought) throws otherwise. it's confusing to me that we're always trying to convert input nodes to constants - only some should be.

eellison commented on torch/csrc/jit/codegen/onednn/graph_helper.cpp:
```diff
@@ -0,0 +1,561 @@
+// that oneDNN executes faster. prim::ListConstruct is an exception, since
+// we simply want to fuse it with cat.
+bool isBetterSuitedForLLGA(NodeKind kindOfOp) {
+  if ((kindOfOp == aten::layer_norm) || (kindOfOp == aten::avg_pool2d) ||
```
nit: `return ((kindOfOp == aten::layer_norm) ...`

eellison reviewed (approved):
Cool, this is looking really good!! i'm accepting. I think we have some follow ups but I think it would more productive to land this and then address the other ones in smaller PRs  than continue to iterate here. Thank you for all your hard work and patience !


follow ups 
- fix test failures with respect to unsupported inputs
- fix up fixConvOptionalBias
- Use opinfo tests
- inferDevice torch_check call
- checkInputCompatibility follow up
- not checking shapes of input mkldnn tensor to llga guard
- the `mayConvertScalarInputToTensor` logic we can consider

(for me) double check `EliminateIdentityMulAdd`

eellison commented:
@sanchitintel mind rebasing and i'll land ?

sanchitintel committed:
Rebase with PyTorch master branch

sanchitintel committed:
Fix style & use setUp & tearDown for UTs

sanchitintel committed:
Fix style

sanchitintel reviewed (commented):
Thanks again for your thorough review, @eellison!
As you advised, we'll address the issues you alluded (today or earlier) to in a subsequent PR.

eellison reviewed (commented):
Hey, getting a build error internally undefiend symbol - `_ZN5torch3jit5fuser6onednn9fuseGraphERSt10shared_ptrINS0_5GraphEE`. Do we need  to add to https://github.com/pytorch/pytorch/blob/master/tools%2Fbuild_variables.bzl?

sanchitintel commented:
Thanks a lot for taking a look, @eellison! To fix this error, we would enable the Bazel build for oneDNN Graph.

sanchitintel committed:
Try to fix non-public Meta Bazel build

malfet commented on caffe2/CMakeLists.txt:
```diff
@@ -629,6 +629,21 @@ if(NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
     include(../mlc/mlc_build.cmake)
   endif()
 
+  if(MKLDNN_FOUND AND NOT WIN32 AND NOT APPLE)
```
Just curious: why it blacklists the change for Windows and Apple? And is it documented somewhere? I.e. if changes are `x86_64` specific they should be guarded as such)

sanchitintel replied:
Currently, we don't support oneDNN Graph integration with PyTorch on Windows & MacOS platforms.
We mentioned that in the [`Limitations` section of the PR description](https://github.com/pytorch/pytorch/pull/68111#issue-1049546420).
However, we can indeed mention it in the `README` file as well. Thanks!

malfet commented on caffe2/CMakeLists.txt:
```diff
@@ -629,6 +629,22 @@ if(NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
 
+  if(MKLDNN_FOUND AND NOT WIN32 AND NOT APPLE)
+    list(APPEND Caffe2_CPU_SRCS
+      ${TORCH_SRC_DIR}/csrc/jit/codegen/onednn/LlgaTensorImpl.cpp
+      ${TORCH_SRC_DIR}/csrc/jit/codegen/onednn/graph_fuser.cpp
+      ${TORCH_SRC_DIR}/csrc/jit/codegen/onednn/graph_rewriter.cpp
+      ${TORCH_SRC_DIR}/csrc/jit/codegen/onednn/graph_helper.cpp
+      ${TORCH_SRC_DIR}/csrc/jit/codegen/onednn/register_interface.cpp
+      ${TORCH_SRC_DIR}/csrc/jit/codegen/onednn/interface.cpp
+      ${TORCH_SRC_DIR}/csrc/jit/codegen/onednn/kernel.cpp
+      ${TORCH_SRC_DIR}/csrc/jit/codegen/onednn/defer_size_check.cpp
+      ${TORCH_SRC_DIR}/csrc/jit/codegen/onednn/layout_propagation.cpp
+      ${TORCH_SRC_DIR}/csrc/jit/codegen/onednn/prepare_binary.cpp
+      ${TORCH_SRC_DIR}/csrc/jit/codegen/onednn/guard_shape.cpp
+    )
```
As @eellison suggested a week ago, why this one is not in `build_variables.bzl`?

sanchitintel replied:
Sorry, we were confused about why the Bazel build was failing internally at Meta, and we would've had to add another Bazel build file for oneDNN Graph, which would've made it harder to pinpoint the failure cause.

malfet commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -6,6 +6,7 @@
 #include <torch/csrc/jit/codegen/cuda/interface.h>
 #include <torch/csrc/jit/codegen/fuser/interface.h>
 #include <torch/csrc/jit/codegen/fuser/kernel_cache.h>
+#include <torch/csrc/jit/codegen/onednn/interface.h>
```
```suggestion
#if !(defined(FBCODE_CAFFE2) || defined(__APPLE__) || defined(_WIN32))
#include <torch/csrc/jit/codegen/onednn/interface.h>
#endif
```

malfet commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -613,6 +614,8 @@ void initJITBindings(PyObject* module) {
             return oldState;
           })
       .def("_jit_nvfuser_enabled", &RegisterCudaFuseGraph::isRegistered)
+      .def("_jit_set_llga_enabled", &RegisterLlgaFuseGraph::setEnabled)
+      .def("_jit_llga_enabled", &RegisterLlgaFuseGraph::isEnabled)
```
```suggestion
#if !(defined(FBCODE_CAFFE2) || defined(__APPLE__) || defined(_WIN32))
      .def("_jit_set_llga_enabled", &RegisterLlgaFuseGraph::setEnabled)
      .def("_jit_llga_enabled", &RegisterLlgaFuseGraph::isEnabled)
 #endif
```

malfet committed:
Do not define llga hooks on unsupported platforms

Should fix linker errors

malfet commented:
And `graph_rewriter.cpp` is full of DOS newlines...

sanchitintel commented on torch/csrc/jit/codegen/onednn/interface.h:
```diff
@@ -0,0 +1,62 @@
+namespace fuser {
+namespace onednn {
+
+static std::atomic<bool> onednn_enabled{false};
+
+std::atomic<bool>& getLlgaEnabled() {
+  return onednn_enabled;
+}
```
Thanks, I'll move it in a subsequent PR.

sanchitintel commented on torch/csrc/jit/codegen/onednn/interface.h:
```diff
@@ -0,0 +1,62 @@
+  return onednn_enabled;
+}
+
+C10_EXPORT void fuseGraph(std::shared_ptr<Graph>& g);
```
If we don't export this symbol, then we get an `undefined symbol` error when we import `torch` in a Python interpreter.

sanchitintel commented on torch/csrc/jit/codegen/onednn/interface.h:
```diff
@@ -0,0 +1,62 @@
+    bool oldState = fuser::onednn::getLlgaEnabled();
+    fuser::onednn::getLlgaEnabled() = enabled;
+    if (enabled) {
+      registerPass(fuser::onednn::fuseGraph);
```
It wouldn't be an issue. We tested it by using `USE_MKLDNN=0` while building from source, but I'll look into it, and will fix it, if required.

sanchitintel commented:
Thanks a ton for your help, @malfet & @eellison! :)

We'll incorporate your suggestions in subsequent PRs.

suo commented:
Looks like this broke master https://hud.pytorch.org/pytorch/pytorch/commit/7dd08230117f4fa8bb82b3524e90fb00340198c7. I am reverting.

suo commented:
@pytorchbot revert this

sanchitintel commented:
> Looks like this broke master https://hud.pytorch.org/pytorch/pytorch/commit/7dd08230117f4fa8bb82b3524e90fb00340198c7. I am reverting.

Oops! Will fix it ASAP.
